File: sdks/java/src/test/java/com/swissarmynoife/sdk/SakMcpClientTest.java
package com.swissarmynoife.sdk;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.OutputStream;
import java.lang.reflect.Type;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class SakMcpClientTest {
  private static final Gson GSON = new Gson();
  private static final Type MAP_TYPE = new TypeToken<Map<String, Object>>() {}.getType();

  @Test
  void pingWithSession() throws Exception {
    HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
    server.createContext(
        "/",
        exchange -> {
          String raw = new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8);
          Map<String, Object> body = GSON.fromJson(raw, MAP_TYPE);
          String method = String.valueOf(body.get("method"));
          switch (method) {
            case "initialize" ->
                writeJson(exchange, 200, "{\"jsonrpc\":\"2.0\",\"id\":1,\"result\":{}}", "sess-java-1");
            case "notifications/initialized" -> {
              try (HttpExchange ignored = exchange) {
                exchange.sendResponseHeaders(202, -1);
              }
            }
            case "tools/call" -> {
              assertEquals("sess-java-1", exchange.getRequestHeaders().getFirst("mcp-session-id"));
              writeJson(
                  exchange,
                  200,
                  "{\"jsonrpc\":\"2.0\",\"id\":2,\"result\":{\"content\":[{\"type\":\"text\",\"text\":\"pong\"}]}}",
                  null);
            }
            default -> {
              try (HttpExchange ignored = exchange) {
                exchange.sendResponseHeaders(500, -1);
              }
            }
          }
        });
    server.start();
    try {
      SakMcpClient mcp = new SakMcpClient("http://127.0.0.1:" + server.getAddress().getPort() + "/");
      assertEquals("pong", mcp.ping());
      assertEquals("sess-java-1", mcp.getSessionId());
    } finally {
      server.stop(0);
    }
  }

  @Test
  void toolsListNoAutoInit() throws Exception {
    HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
    server.createContext(
        "/",
        exchange -> {
          String raw = new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8);
          Map<String, Object> body = GSON.fromJson(raw, MAP_TYPE);
          assertEquals("tools/list", body.get("method"));
          writeJson(exchange, 200, "{\"jsonrpc\":\"2.0\",\"id\":1,\"result\":{\"tools\":[]}}", null);
        });
    server.start();
    try {
      SakMcpClient mcp = new SakMcpClient("http://127.0.0.1:" + server.getAddress().getPort());
      mcp.setAutoInitialize(false);
      @SuppressWarnings("unchecked")
      Map<String, Object> out = (Map<String, Object>) mcp.toolsList();
      assertTrue(out.containsKey("tools"));
    } finally {
      server.stop(0);
    }
  }

  @Test
  void catalogList() throws Exception {
    HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
    server.createContext(
        "/",
        exchange -> {
          String raw = new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8);
          Map<String, Object> body = GSON.fromJson(raw, MAP_TYPE);
          String method = String.valueOf(body.get("method"));
          switch (method) {
            case "initialize" -> writeJson(exchange, 200, "{\"jsonrpc\":\"2.0\",\"id\":1,\"result\":{}}", "s2");
            case "notifications/initialized" -> {
              try (HttpExchange ignored = exchange) {
                exchange.sendResponseHeaders(202, -1);
              }
            }
            default ->
                writeJson(
                    exchange,
                    200,
                    "{\"jsonrpc\":\"2.0\",\"id\":2,\"result\":{\"offers\":[]}}",
                    null);
          }
        });
    server.start();
    try {
      SakMcpClient mcp = new SakMcpClient("http://127.0.0.1:" + server.getAddress().getPort());
      @SuppressWarnings("unchecked")
      Map<String, Object> out = (Map<String, Object>) mcp.catalogList();
      assertTrue(out.containsKey("offers"));
    } finally {
      server.stop(0);
    }
  }

  private static void writeJson(HttpExchange exchange, int code, String json, String session)
      throws IOException {
    if (session != null) {
      exchange.getResponseHeaders().add("mcp-session-id", session);
    }
    exchange.getResponseHeaders().add("Content-Type", "application/json");
    byte[] bytes = json.getBytes(StandardCharsets.UTF_8);
    exchange.sendResponseHeaders(code, bytes.length);
    try (OutputStream os = exchange.getResponseBody()) {
      os.write(bytes);
    }
  }
}

File: sdks/kotlin/src/test/kotlin/com/swissarmynoife/sdk/SakMcpClientTest.kt
package com.swissarmynoife.sdk

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpServer
import java.net.InetSocketAddress
import java.nio.charset.StandardCharsets
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class SakMcpClientTest {
    private val gson = Gson()
    private val mapType = object : TypeToken<Map<String, Any?>>() {}.type

    @Test
    fun pingNegotiatesSession() {
        val server = HttpServer.create(InetSocketAddress("127.0.0.1", 0), 0)
        var n = 0
        server.createContext("/") { exchange ->
            val raw = String(exchange.requestBody.readBytes(), StandardCharsets.UTF_8)
            val body: Map<String, Any?> = gson.fromJson(raw, mapType)
            val method = body["method"]?.toString()
            n++
            when (method) {
                "initialize" ->
                    writeJson(exchange, 200, """{"jsonrpc":"2.0","id":1,"result":{}}""", "sess-kt-1")
                "notifications/initialized" -> {
                    exchange.sendResponseHeaders(202, -1)
                    exchange.close()
                }
                "tools/call" -> {
                    assertEquals("sess-kt-1", exchange.requestHeaders.getFirst("mcp-session-id"))
                    writeJson(
                        exchange,
                        200,
                        """{"jsonrpc":"2.0","id":2,"result":{"content":[{"type":"text","text":"pong"}]}}""",
                        null,
                    )
                }
                else -> {
                    exchange.sendResponseHeaders(500, -1)
                    exchange.close()
                }
            }
        }
        server.start()
        try {
            val mcp = SakMcpClient("http://127.0.0.1:${server.address.port}/")
            assertEquals("pong", mcp.ping())
            assertEquals("sess-kt-1", mcp.sessionId)
            assertTrue(n >= 3)
        } finally {
            server.stop(0)
        }
    }

    @Test
    fun toolsListNoAutoInit() {
        val server = HttpServer.create(InetSocketAddress("127.0.0.1", 0), 0)
        server.createContext("/") { exchange ->
            val raw = String(exchange.requestBody.readBytes(), StandardCharsets.UTF_8)
            val body: Map<String, Any?> = gson.fromJson(raw, mapType)
            assertEquals("tools/list", body["method"])
            writeJson(exchange, 200, """{"jsonrpc":"2.0","id":1,"result":{"tools":[]}}""", null)
        }
        server.start()
        try {
            val mcp =
                SakMcpClient("http://127.0.0.1:${server.address.port}").apply {
                    autoInitialize = false
                }
            @Suppress("UNCHECKED_CAST")
            val out = mcp.toolsList() as Map<String, Any?>
            assertTrue(out.containsKey("tools"))
        } finally {
            server.stop(0)
        }
    }

    @Test
    fun catalogList() {
        val server = HttpServer.create(InetSocketAddress("127.0.0.1", 0), 0)
        server.createContext("/") { exchange ->
            val raw = String(exchange.requestBody.readBytes(), StandardCharsets.UTF_8)
            val body: Map<String, Any?> = gson.fromJson(raw, mapType)
            when (body["method"]?.toString()) {
                "initialize" ->
                    writeJson(exchange, 200, """{"jsonrpc":"2.0","id":1,"result":{}}""", "s2")
                "notifications/initialized" -> {
                    exchange.sendResponseHeaders(202, -1)
                    exchange.close()
                }
                else ->
                    writeJson(exchange, 200, """{"jsonrpc":"2.0","id":2,"result":{"offers":[]}}""", null)
            }
        }
        server.start()
        try {
            val mcp = SakMcpClient("http://127.0.0.1:${server.address.port}")
            @Suppress("UNCHECKED_CAST")
            val out = mcp.catalogList() as Map<String, Any?>
            assertTrue(out.containsKey("offers"))
        } finally {
            server.stop(0)
        }
    }

    private fun writeJson(exchange: HttpExchange, code: Int, json: String, session: String?) {
        if (session != null) {
            exchange.responseHeaders.add("mcp-session-id", session)
        }
        exchange.responseHeaders.add("Content-Type", "application/json")
        val bytes = json.toByteArray(StandardCharsets.UTF_8)
        exchange.sendResponseHeaders(code, bytes.size.toLong())
        exchange.responseBody.use { it.write(bytes) }
    }
}
